docs(tproxy): restructure and expand transparent proxy documentation (#2064)
kumahq/kuma-website: app/_src/networking/transparent-proxy/reachable-services.md

---
title: Reachable Services
content_type: how-to
---

{% assign Kuma = site.mesh_product_name %}
{% assign tproxy = site.data.tproxy %}

> 

Reviewer: Should mention reachableBackends here in case uses use MeshService? 
Btw should we open an issue to warn users when using MeshService exclusive and using reachableServices?

Author: According to https://github.com/kumahq/kuma/issues/10508 reachable services work with MeshServices, so I don't think we need any warnings

> When the transparent proxy is enabled, {{ Kuma }} automatically configures each data plane proxy to connect with every other data plane proxy in the same mesh. This setup ensures broad service-to-service communication but can create issues in large service meshes:

Author: addressed

- **Increased Memory Usage**: The configuration for each data plane proxy can consume significant memory as the mesh grows.
- **Slower Propagation**: Even small configuration changes for one service must be propagated to all data planes, which can be slow and resource-intensive.
- **Excessive Traffic**: Frequent updates generate additional traffic between the control plane and data plane proxies, especially when services are added or changed often.

> In real-world scenarios, services usually need to communicate only with a few other services rather than all services in the mesh. To address these challenges, {{ Kuma }} offers the **Reachable Services** feature, allowing you to define only the services that each proxy should connect to. Specifying reachable services helps reduce memory usage, improves configuration propagation speed, and minimizes unnecessary traffic.

Reviewer: ```suggestion
In real-world scenarios, services usually need to communicate only with a few other services rather than all services in the mesh. To address these challenges, {{ Kuma }} offers the **Reachable Services** feature, allowing you to define only the services that each proxy should connect to. Specifying reachable services avoids the performance issue mentioned above
```

Author: addressed

Here’s how to configure reachable services:

{% tabs reachable-services useUrlFragment=false %}
{% tab reachable-services Kubernetes %}

> Specify the list of reachable services in the `kuma.io/transparent-proxying-reachable-services` annotation, separating each service with a comma. Your workload configuration could look like this:

Reviewer: ```suggestion
Use the annotation `kuma.io/transparent-proxying-reachable-services` with a comma separated list of services your pod connects to. For example:
```

Author: addressed

```yaml
apiVersion: apps/v1
kind: Pod
metadata:
  name: demo-client
  annotations:
    kuma.io/transparent-proxying-reachable-services: "redis_kuma-demo_svc_6379,elastic_kuma-demo_svc_9200"
...
```

> You can update your workload manifests manually or use a command like:

Reviewer: ```suggestion
You can update your pods manually or with kubectl:
```

```sh

> kumactl annotate pods example-app \

Reviewer 2: ```suggestion
kubectl annotate pods example-app \
```

Author: addressed

  "kuma.io/transparent-proxying-reachable-services=redis_kuma-demo_svc_6379,elastic_kuma-demo_svc_9200"
```
{% endtab %}
{% tab reachable-services Universal %}
To specify reachable services in your `Dataplane` resource, add them under the `networking.transparentProxying.reachableServices` path. Here’s an example:

```yaml
type: Dataplane
mesh: default
name: {% raw %}{{ name }}{% endraw %}
networking:
  address: {% raw %}{{ address }}{% endraw %}
  inbound:
    - port: {% raw %}{{ port }}{% endraw %}
      tags:
        kuma.io/service: demo-client
  transparentProxying:
    redirectPortInbound: {{ tproxy.defaults.redirect.inbound.port }}
    redirectPortOutbound: {{ tproxy.defaults.redirect.outbound.port }}
    reachableServices:
      - redis_kuma-demo_svc_6379
      - elastic_kuma-demo_svc_9200 
```
{% endtab %}
{% endtabs %}

> This configuration ensures that `example-app` only connects to `redis_kuma-demo_svc_6379` and `elastic_kuma-demo_svc_9200`, reducing the overhead associated with managing connections to all services in the mesh.

Reviewer: Need to add the hack to reach no services (Use a placeholder like `_noservice_`)

Author: addressed
